### ISSUE TITLE
"skaffold deploy" is not useful

### ISSUE THREAD

**Issue by ahmetb** [GoogleContainerTools/skaffold#922]
I cannot figure out the use case for "skaffold deploy".

This command is not able to populate the image tags. Without that, all it's doing is corrupting the Deployments by deploying the tag-less images.

This command cannot use the result of "skaffold build" like this (as far as I know):

    skaffold build > images.txt
    skaffold deploy < images.txt

Without these functionality this command just appears like a `kubectl apply` that is not compatible with `skaffold run` to me. 

Please consider either making "deploy" cmd work with "build" cmd or consider eliminating to avoid confusing users.

**Comment by balopat** [comment]
The use case would be to use `skaffold deploy` as part of other orchestration tools. I think we should figure out the tagging situation with `skaffold deploy`.

**Comment by mxey** [comment]
See also https://github.com/GoogleContainerTools/skaffold/issues/653

**Comment by cliveza** [comment]
Yeah, I just picked up on this too. I thought I could `

```
RELEASE=v1.0.0 skaffold build
RELEASE=v1.0.0 skaffold deploy
```
While the build command is able to build an image with the correct tag, using an environment variable and envTemplate, the deploy command just ignore this all together.

**Comment by r2d4** [comment]
I think the right fix for this is that `skaffold deploy` should also generate the tag through the git commit tagger (or env tagger).

**Comment by r2d4** [comment]
Eventually we might want to get rid of the "non-stable" taggers that rely on information from the build artifact itself

**Comment by rsanders** [comment]
Could `skaffold build` write the build artifact names to an output file that is also read by `skaffold deploy`?   I see that it has support for outputting those, but it leaves the plumbing as an exercise for the user.

We are using skaffold in our CD pipeline, and we like having "non-stable" components like build ID in the docker image tag.  We also prefer to have separate build -> test -> deploy stages.

**Comment by r2d4** [comment]
@rsanders yes, skaffold build already outputs the go template of the struct that deploy uses. Deploy should be able to read from that generated output

I imagine something like 

```
skaffold build > build.out
# do some work
skaffold deploy --build-artifacts=build.out
```

or

```
skaffold build | skaffold deploy --build-artifacts -
```

or 
```
skaffold build > build.out
# do some work
skaffold deploy --build-artifacts - < build.out
```

**Comment by balopat** [comment]
cc @tejal29

**Comment by FX-HAO** [comment]
I can't agree with this, i use `build` and `deploy` in different pipelines. Sometimes I just need to build the image and then I can decide if we should deploy it.

**Comment by tejal29** [comment]
My Implementation plan right now is 
1. First fix `skaffold build --output` so that we can write the produced build artifact.
2. Support `skaffold deploy` to consume this output
3. Finally add support for git tagger or env tagger for `skaffold build` to tag images built.

**Comment by pepjo** [comment]
Is this issue linked to this one: https://github.com/GoogleContainerTools/skaffold/issues/1352 ?

**Comment by tejal29** [comment]
So with the changes as described [above](https://github.com/GoogleContainerTools/skaffold/issues/922#issuecomment-474710200) this is how the workflow will look like
1. First run skaffold build
```
skaffold build --quiet > build.out
```

( Note: untill we support git tagger or env tagger artifacts listed in build.out with be digest based)
2. Now deploy
```
skaffold deploy --build-artifacts build.out
```
Note: `--build-artifacts` is a new flag to skaffold deploy.

@dgageot  and @balopat, i feel, 
`skaffold build --quiet ` is  unhelpful from user and IDE perspective because, 
- If there was a built error, you only see the error on stdout. The Build log is discarded. Users or even IDE's will have to re-run the command without `--quiet` flag to see at what stage the build failed.
e.g.
```
../../out/skaffold build --quiet > test_error.out
FATA[0000] introducing an error forcefully              
tejaldesai@tejaldesai~/go/src/github.com/GoogleContainerTools/skaffold/examples/getting-started: (pr/tejal29/1841)$ 
```

Maybe we could an explicit flag to build like:
1. `--list-built-artifacts` <file> : make it more explicit that artifacts build will be written to a file.
    If this flag is present, users can still see the `build` output on std in addition to build results written to file.

What do you think?
This could be an additional task i can add once we all agree.

**Comment by pepjo** [comment]
I am trying to understand the need for this when trying to deploy with an image that has already been built:
```
skaffold build --quiet > build.out
```

Our current planned use case is:
1. If someone triggers a pull request `skaffold build` should be triggered in CI (creating and pushing an image with the git commit as the tag)
2. When the pull request is merged CI would run `skaffold deploy` to deploy the image that we already created. It will use the same naming convention to try to deploy the image that corresponds to the commit. This only works if the merges are fast-forward as the git commit reference will be kept. A useful addition would be to add a flag that checks with the docker registry if the image does indeed exist and return 1 if not.

Adding the `build.out` is complicated (and unnecessary?) in this use case. We do continuous deployment so it's ok for us to not have meaningful docker tags (eg. with version numbers).

Will this use case be supported with the planned implementation?

**Comment by tejal29** [comment]
>Adding the build.out is complicated (and unnecessary?) in this use case. We do continuous deployment so it's ok for us to not have meaningful docker tags (eg. with version numbers).

The first thing we are going to address is deploy should be able to consume build output and its easier if we first implement a `build.out` based approach.

> Will this use case be supported with the planned implementation?
@pepjo we mention your use case in follow up ideas where `skaffold deploy` can infer tags from git or an env variable.

**Comment by pepjo** [comment]
Thanks, I did not see that particular part of the comment. Thanks for the answer.

**Comment by tejal29** [comment]
#1986  fixes one of the subtasks for "Verify #1352 still exists and if it does, do not run skaffold build in skaffold deploy"
